File: Leetcode/dfs_recursion_backtracking/563_BinaryTreeTilt/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findTilt(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        res = [0]
        self.dfs(root, res)
        return res[0]

    def dfs(self, root, res):
        if not root: return 0
        left = self.dfs(root.left, res)
        right = self.dfs(root.right, res)
        res[0] += abs(left-right)
        return root.val+left+right

File: Leetcode/dfs_recursion_backtracking/563_BinaryTreeTilt/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    int dfs(TreeNode* root, int& res){
        if(root==nullptr) return 0;
        int left = dfs(root->left, res);
        int right = dfs(root->right, res);
        res += abs(left-right);
        return root->val+left+right;
    }
public:
    int findTilt(TreeNode* root) {
        int res = 0;
        dfs(root, res);
        return res;
    }
};
